File: license_api/src/routes/license.ts
import { Hono } from "hono";
import { zValidator } from "@hono/zod-validator";
import { z } from "zod";
import {
  validateLicenseKey,
  activateLicense,
  deactivateLicense,
  getLicenseById,
} from "../polar.ts";

const licenseRoutes = new Hono();

// Validate license key
const validateSchema = z.object({
  licenseKey: z.string().min(1, "License key is required"),
});

licenseRoutes.post("/validate", zValidator("json", validateSchema), async (c) => {
  const { licenseKey } = c.req.valid("json");

  const result = await validateLicenseKey(licenseKey);

  if (!result.valid) {
    return c.json(
      {
        valid: false,
        error: result.error ?? "Invalid license key",
      },
      400
    );
  }

  return c.json({
    valid: true,
    license: result.license,
  });
});

// Activate license
const activateSchema = z.object({
  licenseKey: z.string().min(1, "License key is required"),
  deviceId: z.string().min(1, "Device ID is required"),
  deviceName: z.string().optional(),
  meta: z.record(z.string(), z.string()).optional(),
});

licenseRoutes.post("/activate", zValidator("json", activateSchema), async (c) => {
  const { licenseKey, deviceId, deviceName, meta } = c.req.valid("json");

  // First validate the license
  const validation = await validateLicenseKey(licenseKey);
  if (!validation.valid) {
    return c.json(
      {
        success: false,
        error: validation.error ?? "Invalid license key",
      },
      400
    );
  }

  // Activate
  const result = await activateLicense(licenseKey, deviceName ?? deviceId, {
    deviceId,
    ...meta,
  });

  if (!result.success) {
    return c.json(
      {
        success: false,
        error: result.error,
      },
      400
    );
  }

  return c.json({
    success: true,
    activationId: result.activationId,
    license: validation.license,
  });
});

// Deactivate license
const deactivateSchema = z.object({
  licenseKey: z.string().min(1, "License key is required"),
  activationId: z.string().min(1, "Activation ID is required"),
});

licenseRoutes.post("/deactivate", zValidator("json", deactivateSchema), async (c) => {
  const { licenseKey, activationId } = c.req.valid("json");

  const result = await deactivateLicense(licenseKey, activationId);

  if (!result.success) {
    return c.json(
      {
        success: false,
        error: result.error,
      },
      400
    );
  }

  return c.json({
    success: true,
  });
});

// Get license by ID (for admin purposes)
const getLicenseSchema = z.object({
  id: z.string().min(1, "License ID is required"),
});

licenseRoutes.get("/:id", zValidator("param", getLicenseSchema), async (c) => {
  const { id } = c.req.valid("param");

  const license = await getLicenseById(id);

  if (!license) {
    return c.json(
      {
        error: "License not found",
      },
      404
    );
  }

  return c.json({
    license,
  });
});

export { licenseRoutes };

File: src-tauri/src/license.rs
use parking_lot::RwLock;
use reqwest::Client;
use serde::{Deserialize, Serialize};
use std::sync::Arc;

const LICENSE_API_URL: &str = "https://api.querystudio.dev";

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LicenseInfo {
    pub id: String,
    pub key: String,
    pub display_key: String,
    pub status: String,
    pub customer_id: String,
    pub email: String,
    pub customer_name: Option<String>,
    pub activations_count: i32,
    pub max_activations: Option<i32>,
    pub usage: i32,
    pub max_usage: Option<i32>,
    pub validations: i32,
    pub expires_at: Option<String>,
    pub created_at: String,
    pub benefit_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ValidateLicenseResponse {
    pub valid: bool,
    pub license: Option<LicenseInfo>,
    pub error: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ActivateLicenseResponse {
    pub success: bool,
    pub activation_id: Option<String>,
    pub license: Option<LicenseInfo>,
    pub error: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DeactivateLicenseResponse {
    pub success: bool,
    pub error: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StoredLicense {
    pub license_key: String,
    pub activation_id: String,
    pub device_id: String,
    pub license_info: LicenseInfo,
    pub validated_at: String,
}

pub struct LicenseManager {
    client: Client,
    license: RwLock<Option<StoredLicense>>,
}

impl LicenseManager {
    pub fn new() -> Self {
        Self {
            client: Client::new(),
            license: RwLock::new(None),
        }
    }

    pub fn get_device_id() -> String {
        machine_uid::get().unwrap_or_else(|_| uuid::Uuid::new_v4().to_string())
    }

    pub fn is_licensed(&self) -> bool {
        self.license.read().is_some()
    }

    pub fn get_license(&self) -> Option<StoredLicense> {
        self.license.read().clone()
    }

    pub fn set_license(&self, license: StoredLicense) {
        *self.license.write() = Some(license);
    }

    pub fn clear_license(&self) {
        *self.license.write() = None;
    }

    pub async fn validate_license(&self, license_key: &str) -> Result<ValidateLicenseResponse, String> {
        let url = format!("{}/api/license/validate", LICENSE_API_URL);

        let response = self
            .client
            .post(&url)
            .json(&serde_json::json!({ "licenseKey": license_key }))
            .send()
            .await
            .map_err(|e| format!("Failed to connect to license server: {}", e))?;

        if response.status().is_success() {
            response
                .json::<ValidateLicenseResponse>()
                .await
                .map_err(|e| format!("Failed to parse response: {}", e))
        } else {
            let error_body = response.text().await.unwrap_or_default();
            Err(format!("License validation failed: {}", error_body))
        }
    }

    pub async fn activate_license(
        &self,
        license_key: &str,
        device_name: Option<&str>,
    ) -> Result<ActivateLicenseResponse, String> {
        let device_id = Self::get_device_id();
        let url = format!("{}/api/license/activate", LICENSE_API_URL);

        let mut body = serde_json::json!({
            "licenseKey": license_key,
            "deviceId": device_id,
        });

        if let Some(name) = device_name {
            body["deviceName"] = serde_json::Value::String(name.to_string());
        }

        let response = self
            .client
            .post(&url)
            .json(&body)
            .send()
            .await
            .map_err(|e| format!("Failed to connect to license server: {}", e))?;

        if response.status().is_success() {
            let result = response
                .json::<ActivateLicenseResponse>()
                .await
                .map_err(|e| format!("Failed to parse response: {}", e))?;

            // Store the license if activation was successful
            if result.success {
                if let (Some(activation_id), Some(license_info)) =
                    (result.activation_id.clone(), result.license.clone())
                {
                    let stored = StoredLicense {
                        license_key: license_key.to_string(),
                        activation_id,
                        device_id,
                        license_info,
                        validated_at: chrono::Utc::now().to_rfc3339(),
                    };
                    self.set_license(stored);
                }
            }

            Ok(result)
        } else {
            let error_body = response.text().await.unwrap_or_default();
            Err(format!("License activation failed: {}", error_body))
        }
    }

    pub async fn deactivate_license(&self) -> Result<DeactivateLicenseResponse, String> {
        let stored = self
            .get_license()
            .ok_or_else(|| "No active license".to_string())?;

        let url = format!("{}/api/license/deactivate", LICENSE_API_URL);

        let response = self
            .client
            .post(&url)
            .json(&serde_json::json!({
                "licenseKey": stored.license_key,
                "activationId": stored.activation_id,
            }))
            .send()
            .await
            .map_err(|e| format!("Failed to connect to license server: {}", e))?;

        if response.status().is_success() {
            let result = response
                .json::<DeactivateLicenseResponse>()
                .await
                .map_err(|e| format!("Failed to parse response: {}", e))?;

            if result.success {
                self.clear_license();
            }

            Ok(result)
        } else {
            let error_body = response.text().await.unwrap_or_default();
            Err(format!("License deactivation failed: {}", error_body))
        }
    }

    pub async fn revalidate_license(&self) -> Result<bool, String> {
        let stored = match self.get_license() {
            Some(s) => s,
            None => return Ok(false),
        };

        let result = self.validate_license(&stored.license_key).await?;

        if !result.valid {
            self.clear_license();
            return Ok(false);
        }

        // Update license info
        if let Some(license_info) = result.license {
            let updated = StoredLicense {
                license_info,
                validated_at: chrono::Utc::now().to_rfc3339(),
                ..stored
            };
            self.set_license(updated);
        }

        Ok(true)
    }
}

pub type LicenseState = Arc<LicenseManager>;
